File: trees/bstree.py
import random

class BinaryTreeNode:
  """Binary Tree Node"""
  def __init__(self, key, left=None, right=None):
    self.key = key
    self.left = left
    self.right = right

class BinarySearchTree:
  """Binary Search Tree"""
  def __init__(self, root=None):
    self.root = root

  def insert(self, key):
    success, self.root = self._insert(self.root, key)
    if success:
      print 'key', key, 'inserted into the bst'
    else:
      print 'key', key, 'already present'

  def _insert(self, node, key):
    if node is None:
      return True, BinaryTreeNode(key)
    if node.key == key:
      return False, node
    else:
      if node.key > key:
        success, node.left = self._insert(node.left, key)
      else:
        success, node.right = self._insert(node.right, key)
      return success, node
  
  def get_successor(self, node):
    if node.right is None:
      return None
    node = node.right
    while node.left is not None:
      node = node.left
    return node

  def remove(self, key):
    success, self.root = self._remove(self.root, key)
    if success:
      print 'key', key, 'removed from the bst'
    else:
      print 'key', key, 'not present'

  def _remove(self, node, key):
    if node is None:
      return False, node
    if node.key == key:
      if node.left is not None and node.right is not None:
        successor = self.get_successor(node)
        node.key = successor.key
        success, node.right = self._remove(node.right, successor.key)
        return True, node
      elif node.left is not None:
        return True, node.left
      else:
        return True, node.right
    else:
      if node.key > key:
        success, node.left = self._remove(node.left, key)
      else:
        success, node.right = self._remove(node.right, key)
      return success, node

  def inorder(self):
    if self.root is None:
      print 'Empty tree !'
    else:
      self._inorder(self.root)
      print

  def _inorder(self, node):
    if node is not None:
      self._inorder(node.left)
      print node.key,
      self._inorder(node.right)
   

if __name__ == '__main__':
  tree = BinarySearchTree();
  tree.inorder()
  n = 50
  for i in xrange(n):
    tree.insert(random.randint(1, n))
    tree.inorder()
  for i in xrange(n*n):
    tree.remove(random.randint(1, n))
    tree.inorder()

File: trees/BSTree.java
import java.util.Comparator;
import java.util.Random;
import java.util.ArrayList;

class Node<T extends Comparable<T>> {
  T key;
  Node<T> left, right;

  Node() {
    this.left = null;
    this.right = null;
  }

  Node(T x) {
    super();
    this.key = x;
  }
}

class BSTree<T extends Comparable<T>> {
  Node<T> root;
  Comparator<T> comparator;

  BSTree() {
    this.comparator = null;
    this.root = null;
  }

  BSTree(Comparator<T> comparator) {
    this.comparator = comparator;
    this.root = null;
  }

  private int compare(T x, T y) {
    if (comparator == null)
      return x.compareTo(y);
    else
      return comparator.compare(x, y);
  }

  void insert(T key) {
    root = insert(root, key);
  }

  private Node<T> insert(Node<T> node, T key) {
    if (node == null) {
      return new Node<T>(key);
    }
    else if (compare(key, node.key) < 0) {
      node.left = insert(node.left, key);
    }
    else if (compare(key, node.key) > 0) {
      node.right = insert(node.right, key);
    }
    else {
      System.err.println("Element " + key + " is already in the BST !");
    }
    return node;
  }

  void remove(T key) {
    this.root = remove(this.root, key);
  }

  private Node<T> remove(Node<T> node, T key) {
    if (node == null) {
      System.err.println("Element " + key + " is not present in the BST !");
    }
    else if (compare(key, node.key) < 0) {
      node.left = remove(node.left, key);
    }
    else if (compare(key, node.key) > 0) {
      node.right = remove(node.right, key);
    }
    else {
      if (node.left == null)        node = node.right;
      else if (node.right == null)  node = node.left;
      else {
        T rightSuccessorKey = getMin(node.right);
        node.key = rightSuccessorKey;
        node.right = remove(node.right, rightSuccessorKey);
      }
    }
    return node;
  }

  private T getMin(Node<T> node) {
    if (node == null)   return null;
    while (node.left != null) {
      node = node.left;
    }
    return node.key;
  }

  void printInOrder() {
    if (root == null) {
      System.err.println("Tree is empty !");
    }
    else {
      printInOrder(root);
      System.out.println();
    }
  }

  private void printInOrder(Node<T> node) {
    if (node == null)   return;
    printInOrder(node.left);
    System.out.print(" " + node.key);
    printInOrder(node.right);
  }

  public static void main (String [] args) {
    BSTree<Integer> bst = new BSTree<Integer>();

    Random random = new Random();

    int numNodes = 10;
    System.out.println("Number of nodes = " + numNodes);

    ArrayList<Integer> list = new ArrayList<Integer>();
    boolean[] deleted = new boolean[numNodes];

    for (int i = 0; i < numNodes; i++) {
      int num = random.nextInt(1000);
      list.add(num);
      deleted[i] = false;

      System.out.println("Element to be inserted: " + num);
      bst.insert(num);
      bst.printInOrder();
    }

    for(int i = 0; i < numNodes; i++) {
      int index = random.nextInt(numNodes);
      while (deleted[index]) {
        index = random.nextInt(numNodes);
      }

      deleted[index] = true;

      int num = list.get(index);

      System.out.println("Element to be deleted: " + num);
      bst.remove(num);
      bst.printInOrder();
    }
  }
}
